README.md

```diff
@@ -418,6 +418,62 @@ logback.xml Example:
     </root>
 </configuration>
 ```
+Kafka Sender configuration
+--------------------------
+If you want to use Kafka as transportation input you should define at least host(with schema kafka), port and kafkaLogTopic property
+
+Sample log4J2 configuration:
+```xml
+<Configuration packages="biz.paluch.logging.gelf.log4j2">
+    <Appenders>
+        <Gelf name="gelf" host="kafka:localhost" port="9092" version="1.1" extractStackTrace="true"
+              filterStackTrace="true" mdcProfiling="true" includeFullMdc="true" maximumMessageSize="8192"
+              originHost="%host{fqdn}" additionalFieldTypes="fieldName1=String,fieldName2=Double,fieldName3=Long"
+              kafkaLogTopic="kafka-log-topic" kafkaProducerProperties="/custompath/producer.properties">
```

Comment:
we should not extend general properties for a specific sender implementation. We should rather move additional properties into the `host` property (see Redis implementation ([Documentation](https://logging.paluch.biz/redis.html) for further details).

Reply:
Well actually I started same as that. But while implementing I saw that Kafka Producer has very complex configuration options especially about security. 
Please see(https://docs.confluent.io/current/installation/configuration/producer-configs.html) 
So I decided to provide it on KafkaProducerProperties property for complex multibroker implementations. Even it works without providing that. 
It needs only kafkaLogTopic with host and port of a broker for PLAINTEXT listener configuration of a single broker. I first thought it can also be moved to a sementic like kafka:localhost#kafka-log-topic but about usability i changed it that way.
And if it is not a kafkasender these configurations is not requested for any place so can be null.
If we move them to addiotional fields these will be logged to gray log so again we need to remove them at runtime?

Reply:
What kind of semantic should it be? If it is going to be defined inside host. Because it has multibroker initilization. Like 192.168.1.30:9202,192.168.1.31:9093,etc.
Is that going to be a must sinle broker? @mp911de

Reply:
I'm not familiar with Kafka and the settings that are typically configured. The purpose of logstash-gelf is to solve a common problem for the majority of users and this requires us to keep complexity at a minimal level. If you need to configure a lot of properties for your Kafka integration, then implementing a custom `GelfSenderProvider` should be the way to go.

Could we have something like `kafka:localhost:54759?request.timeout.ms=500&…#kafka-log-topic`/`kafkas:localhost:54759#kafka-log-topic` or do you think many other properties are required?

Reply:
Re multi-broker: How about `kafka:host1:port1,host2:port2?option=value#topic`?

Reply:
Hımm. Actually we are getting out of standarts of other implementations in tcp, udp, etc. Like host, port definitions also leaving url standarts. But you are the boss. I can write a uri parser for that if you want it.

Reply:
Ok. I'm implementing like that. @mp911de

Reply:
Oh it can not. It has properties like ssl keystore location inside config. Please take a look:https://docs.hortonworks.com/HDPDocuments/HDP2/HDP-2.6.5/bk_security/content/ch_wire-kafka.html
@mp911de

Reply:
We should not require SSL properties as SSL support should default to JVM settings (`-Djavax.net.ssl.keyStore=…`).

I agree there is a mismatch between `host`, `port` and multi-broker usage. This library was written with TCP and UDP support in mind. We added HTTP and Redis support over time – each transport requires a different set of properties. So either we represent additional properties in the host field or we consider a transport not a good fit for logstash-gelf.

Reply:
If we are ok with nonstandard semantics it may be like pipe separator with |  like:

> kafka:[brokers]**|**[topic]**|** [options_as_query_parameters]
```
kafka:192.168.1.30:9092,192.168.1.31:9093|kafka-log-topic|acks="all"&ssl.keystore.location ="/var/private/ssl/kafka.server.keystore.jks"
```
I'm changing implementation this way.

Reply:
And port will not be important for kafka configuration. Is that ok? @mp911de

Reply:
> We should not require SSL properties as SSL support should default to JVM settings (`-Djavax.net.ssl.keyStore=…`).
> 
I gave that as a sample for file path property. Actually I don't see such property inside as static definition (https://kafka.apache.org/21/javadoc/?org/apache/kafka/clients/producer/ProducerConfig.html) but exists in http://kafka.apache.org/documentation.html#producerconfigs

Reply:
`javax.net.ssl.*` properties come from Java's SSL/TLS support and configure the default `SSLContext`.

Reply:
@mp911de Do you want me to remove all properties using `javax.net.ssl.*` inside creation if specified inside options_as_query_parameters?

Reply:
> And port will not be important

That's fine.

>  we are ok with nonstandard semantics it may be like pipe separator with | like:

Why? `bootstrap.servers` uses as well comma notation and I think we should stick to what is already common when configuring Kafka connectivity.

> Do you want me to remove all properties using javax.net.ssl.* inside 

Currently, we should not filter (remove) properties unless they cause issues.

Reply:
I'm sorry for my lack of knowledge about Uri. I did not know that we can get comma-separated values with `URI.getAuthority()`.  I realized that after your comment down:
> We should convert the `host` value into an `URI` to get some assistance from existing implementations. See [here](https://github.com/lettuce-io/lettuce-core/blob/master/src/main/java/io/lettuce/core/RedisURI.java#L903) for an example how to parse comma-separated host values.

Also about my not being careful about your guidence here:
> we should not extend general properties for a specific sender implementation. We should rather move additional properties into the `host` property (see Redis implementation ([Documentation](https://logging.paluch.biz/redis.html) for further details).

In which already implemented same thing with redis-sentinental uri. Changing sementic implementation to:
> kafka://bootsrap_servers?options_as_query_parameters#topic

Now it is clear. 😄

Reply:
All good. Pull requests are for discussing, sharing knowledge and collaboration. Happy to guide you through all the things.